wordpress-mobile/WordPress-Android
# Add network mocking for end to end tests

This PR adds network mocking for all of our end-to-end (Espresso) tests, including screenshot tests.

- This uses [WireMock](http://wiremock.org/) to power mocking the network requests. you can see the docs [here](http://wiremock.org/docs/).
- WireMock just runs a small HTTP server on `localhost:8080` which mimics the WP API. Request/response configuration is done with the JSON files. I hope to use the same JSON files to mock the iOS tests.
- This URL is injected into FluxC in place of `public-api.wordpress.com`. Thanks @aforcier for implementing this.
- Images are not currently mocked, and still come from the web. This was proving tricky and I didn't want to block on it any longer. I will follow up on this later.
- There is a new subproject, `WordPressMocks` which contains all of the configuration and mock files for WireMock. We can make this a subtree later and share the mocks with iOS.

The diff is very big, but it should be not too difficult to review as it is mostly JSON files 😄 You can see it without the JSON files [here](https://github.com/wordpress-mobile/WordPress-Android/pull/9422/files?file-filters%5B%5D=.bat&file-filters%5B%5D=.gradle&file-filters%5B%5D=.jar&file-filters%5B%5D=.java&file-filters%5B%5D=.kt&file-filters%5B%5D=.properties&file-filters%5B%5D=.properties-example&file-filters%5B%5D=.xml&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile) for easier reviewing. Without the JSON files, this is the diff size 😄 

```
27 files changed, 706 insertions(+), 80 deletions(-)
```

Still to do:

- [x] Add support for the screenshot tests once https://github.com/wordpress-mobile/WordPress-Android/pull/9368 is merged.

~Once the above is completed, this will be ready for review.~

I plan to move `libs/mocks` to a subtreed repo so the mocks defined here can be shared with iOS.

To test:

1. To esnure you are not using live secrets, and everything is mocked: `cp gradle.properties-example gradle.properties`.
2. Run `./gradlew WordPress:connectedVanillaDebugAndroidTest --stacktrace`.
3. All the tests should pass.

Update release notes:
